add sd adapter for converting unsupported SD type to filesd

cstyan opened the pull request:
Still in progress, for #3658

So far I've just been using filesd as input for the tool, but I'm writing a quick example SD implementation and adding tests still.

Reuses `Manager` to dedup and maintain separate target groups in the same way that users SD implementation will produce them.

krasi-georgiev commented on discovery/manager.go:
```diff
@@ -115,6 +115,13 @@ func (m *Manager) ApplyConfig(cfg map[string]sd_config.ServiceDiscoveryConfig) e
+	poolKey := poolKey{setName: name, provider: name}
+	m.startProvider(ctx, poolKey, worker)
+}
+
```
is there no way to do this without modifying the Discovery manager?

cstyan replied:
Not that I can see. SD providers are normally started via `providersFromConfig` which only knows about existing SD types. Also, I think it's best if knowledge of `poolKey` is contained to `Manager`.

krasi-georgiev replied:
I will try to refactor the SD a bit to allow its usage without such modification.
I am thinking to implement the Discoverers as plugable middleware so that this can be done at at the initialisation.

krasi-georgiev commented:
https://github.com/prometheus/prometheus/pull/3765
@cstyan this will make using the Discovery package very easy and can register custom Discoverers.
Please try it and let me know what you think.

cstyan commented:
@krasi-georgiev the issue with that is that now you also need to make changes to `Discovery.Config`, since those functions only handle existing service discovery implementations `SDConfig` types. I don't see how that plus the other changes in your PR are any simpler than one small change to `Manager`.

krasi-georgiev commented on discovery/file/adapter/adapter.go:
```diff
@@ -0,0 +1,109 @@
+}
+
+// Adapter runs an unknown service discovery implementation and converts it's target groups
+// to json and writes to a file for file_sd.
```
how about using `custom` instead of unknown?

the same for all other places.

krasi-georgiev commented on documentation/examples/sd-adapter/adapter-usage/main.go:
```diff
@@ -0,0 +1,138 @@
+		tg.Targets = targetsLabels
+		tgs = append(tgs[:], &tg)
+	}
+	fmt.Printf("end tgs: %+v\n", tgs)
```
can't we use a real logger here?

krasi-georgiev replied:
the same for the other places.

cstyan commented:
@krasi-georgiev updated most of what we talked about, not sure what each file is in the renaming you proposed. I think `adapter.go` should remain where it is now. I updated the readme in the example in case it was unclear what each file was.

krasi-georgiev commented:
```
documentation/examples/customDiscovery/server/main.go // the mock server that servers the SD targets
documentation/examples/customDiscovery/provider.go //  the Discoverer that implements the Discoverer interface
documentation/examples/customDiscovery/main.go // main entry that glues all together.
documentation/examples/customDiscovery/sdFileWriter.go // the adapter that writes the json file
documentation/examples/customDiscovery/README,md
```

Why do you think that the adapter should be separated?
I think the example should be self contained.

cstyan commented:
The adapter is something that we should maintain and have some sort of guarantee to users that it works, which isn't necessarily the case if it's in `documentation/examples`, at least that's my thinking. The example is strictly an example of usage of the adapter and writing a `Discoverer`.

Does that make sense?

krasi-georgiev commented:
I am not sure which would be better.
@Gouthamve do you think the tool the writes the json file should be part of the examples folder or as a package inside the SD file folder?

cstyan commented:
Yeah, it would be nice to get some feedback from others who have worked on the SD related code :)

brian-brazil commented:
Examples is where it belongs, but we can hook that (and the adapter) into CI.

cstyan commented:
I think the question was whether the adapter code itself should live in the
discovery package or in the examples.

On Wednesday, February 7, 2018, Brian Brazil <notifications@github.com>
wrote:

> Examples is where it belongs, but we can hook that (and the adapter) into
> CI.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/prometheus/prometheus/pull/3720#issuecomment-364027460>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADGJnP3JDw3Wr85mL9p_yuMGQfKtlKrSks5tSqTxgaJpZM4Rl_N2>
> .
>

krasi-georgiev commented:
`/custom-sd/adapter-usage/main.go` --> `/custom-sd/main.go` what do you think ?

it will be in the same folder as the README.md

cstyan commented:
@krasi-georgiev I don't feel strongly either way, wherever the files are I'll just make sure that's reflected in the readme.

cstyan commented:
@brian-brazil @krasi-georgiev this is done imo, I updated the readme and I don't think tests are necessary since this is all example code.

brian-brazil commented on documentation/examples/custom-sd/adapter-usage/main.go:
```diff
@@ -0,0 +1,155 @@
+)
+
+var (
+	a             = kingpin.New("sd adapter usage", "Tool to generate file_sd config files for unimplemented SD mechanisms.")
```
They're files with targets more than config files

brian-brazil commented on documentation/examples/custom-sd/adapter-usage/main.go:
```diff
@@ -0,0 +1,155 @@
+
+var (
+	a             = kingpin.New("sd adapter usage", "Tool to generate file_sd config files for unimplemented SD mechanisms.")
+	outputFile    = a.Flag("output.file", "Output file for file_sd compatible file. Prints to stdout if none provided.").Default("/tmp/custom_sd.json").String()
```
Use the current directory rather than /tmp

brian-brazil replied:
Providing a default and also saying "Prints to stdout if none provided." is confusing.

brian-brazil commented on documentation/examples/custom-sd/adapter-usage/main.go:
```diff
@@ -0,0 +1,155 @@
+var (
+	a             = kingpin.New("sd adapter usage", "Tool to generate file_sd config files for unimplemented SD mechanisms.")
+	outputFile    = a.Flag("output.file", "Output file for file_sd compatible file. Prints to stdout if none provided.").Default("/tmp/custom_sd.json").String()
+	listenAddress = a.Flag("listen.address", "The address the http sd is listening on for requests.").String()
```
HTTP

brian-brazil replied:
This should have a default

brian-brazil commented on documentation/examples/custom-sd/adapter-usage/main.go:
```diff
@@ -0,0 +1,155 @@
+	a             = kingpin.New("sd adapter usage", "Tool to generate file_sd config files for unimplemented SD mechanisms.")
+	outputFile    = a.Flag("output.file", "Output file for file_sd compatible file. Prints to stdout if none provided.").Default("/tmp/custom_sd.json").String()
+	listenAddress = a.Flag("listen.address", "The address the http sd is listening on for requests.").String()
+	logFile       = a.Flag("log.file", "Location of logfile.").String()
```
This is not our usual way of doing logs

brian-brazil commented on documentation/examples/custom-sd/adapter-usage/main.go:
```diff
@@ -0,0 +1,155 @@
+func (d *discovery) Run(ctx context.Context, ch chan<- []*targetgroup.Group) {
+	targetsURL := fmt.Sprintf("http://%s/targets", d.listenAddress)
+
+	// send the target groups down the channel
```
Comments should be full sentances

brian-brazil commented on documentation/examples/custom-sd/adapter-usage/main.go:
```diff
@@ -0,0 +1,155 @@
+		return nil, err
+	}
+
+	targetsMap := arbJSON.(map[string]interface{})
```
This will panic on bad data

I'd suggest using normal JSON parsing rather than walking the struct by hand.

cstyan replied:
This is the recommended way to decode arbitrary json: https://blog.golang.org/json-and-go

I just wrote the example HTTP SD to return json like this:
```
{"euw":["database"],"use":["database","webserver"]}
```
Here each field is some arbitrary datacenter name with an array of the service names running there. So if we want to `Unmarshal` into an actual struct and not have a type assertion I'll change the structure of the data that the HTTP SD returns. But that depends on how strongly you feel this needs to change. 

The alternative is that I just change that line to `targetsMap, ok := arbJSON.(map[string]interface{})`. This will remove to possibility of a panic and we can check the value of `ok` and return if we couldn't assert to the right type.

brian-brazil replied:
All the exporters I know of use standard json unmarshalling, let's not confuse the matter.

brian-brazil commented on documentation/examples/custom-sd/adapter/adapter.go:
```diff
@@ -0,0 +1,118 @@
+	Labels  map[string]string `json:"labels"`
+}
+
+// Adapter runs an unknown service discovery implementation and converts it's target groups
```
its

brian-brazil commented on documentation/examples/custom-sd/adapter/adapter.go:
```diff
@@ -0,0 +1,118 @@
+}
+
+// Adapter runs an unknown service discovery implementation and converts it's target groups
+// to json and writes to a file for file_sd.
```
JSON

brian-brazil commented on documentation/examples/custom-sd/adapter/adapter.go:
```diff
@@ -0,0 +1,118 @@
+		a.groups = tempGroups
+		arr := mapToArray(a.groups)
+		b, _ := json.MarshalIndent(arr, "", "    ")
+		ioutil.WriteFile(a.output, b, 0644)
```
I would not expect a method called "parse" to write to files

brian-brazil replied:
This needs a rename() to be atomic

brian-brazil commented on documentation/examples/custom-sd/adapter-usage/main.go:
```diff
@@ -0,0 +1,155 @@
+	}
+}
+
+func main() {
```
Why do you have two main functions?

cstyan replied:
I wrote the example as two executables, `adapter-usage/main.go` is the executable that queries the SD implementation and turns it's target groups into JSON to write for file_sd, `http-discovery/main.go` is the executable that runs this mock HTTP SD.

cstyan commented:
Thanks for the review @brian-brazil I addressed most your comments and left replies for others.

Regarding `It's not clear to me from a first glance how I as an SD implementer would use this and what I do/don't have to change`, would you prefer the README be more clear, or we change the folder structure so potential implementers hopefully won't have to look at the README to figure out how to use/modify the example.

krasi-georgiev commented:
@cstyan can you please add  to the description that when this gets merged it will close.
https://github.com/prometheus/prometheus/pull/2883
https://github.com/prometheus/prometheus/pull/2221

krasi-georgiev commented:
also closes https://github.com/prometheus/prometheus/pull/3369

brian-brazil commented:
Let's not auto-close those, a more custom message for each will be better received.

> would you prefer the README be more clear, or we change the folder structure so potential implementers hopefully won't have to look at the README to figure out how to use/modify the example.

I'd make the readme clearer, and comments around the bits that the user needs to provide/change.

I'm also wondering if this separate SD invented here is a good idea, I can see it causing confusion. Maybe do a very rudimentary Consul SD?

krasi-georgiev commented:
do you mean instead of the custom http SD server use consul  to serve the targets? 
I think this might cause even more confusion.

maybe simplify the  http server, but keep the general idea.
can't we just have  `/targets` and `/targets?filter=updates`
just to show that  initially we need all targets and than can process only the updates. 

the target source is used as the identifier.
https://github.com/prometheus/prometheus/blob/f0fd70130932bf6bcce2a7877524924170f1e9b0/discovery/manager.go#L183-L190

brian-brazil commented:
> do you mean instead of the custom http SD server use consul to serve the targets?

Yes. I can see users thinking that this is part of how SD works in Prometheus, rather than something to facilitate the example.

krasi-georgiev commented:
I don't mind either way, the general idea is to be easy to understand and easy to implement.

cstyan commented:
> Let's not auto-close those, a more custom message for each will be better received.

Agreed, we can just leave a message on each issue letting them know the adapter now exists.

> I'm also wondering if this separate SD invented here is a good idea, I can see it causing confusion. Maybe do a very rudimentary Consul SD?

I just wanted an example that showed that the adapter is used for possible SD mechanisms that aren't yet in Prometheus, and that in this case the user has written (or written a wrap around something) themselves. I think another implementation of a SD mechanism we already support could be confusing.

I've pushed a commit with comments and updates to the readme so that hopefully it's clearer what someone wanting to write a new SD implementation would have to do.

brian-brazil commented on documentation/examples/custom-sd/adapter-usage/main.go:
```diff
@@ -0,0 +1,156 @@
+	a             = kingpin.New("sd adapter usage", "Tool to generate file_sd target files for unimplemented SD mechanisms.")
+	outputFile    = a.Flag("output.file", "Output file for file_sd compatible file.").Default("/tmp/custom_sd.json").String()
+	listenAddress = a.Flag("listen.address", "The address the HTTP sd is listening on for requests.").Default("localhost:8080").String()
+	logFile       = a.Flag("log.file", "Location of logfile.").String()
```
This is still non-standard

brian-brazil commented on documentation/examples/custom-sd/adapter-usage/main.go:
```diff
@@ -0,0 +1,156 @@
+	return tgs, nil
+}
+
+// Note: you must implement this function for your discovery implementation as part of the
```
What does this function need to do?

brian-brazil commented on documentation/examples/custom-sd/adapter-usage/main.go:
```diff
@@ -0,0 +1,156 @@
+// ex: {"euw":["webserver"],"hq-test":["webserver","db"],"use":["webserver"]}
+// We need to convert this JSON to our internal type targetgroup.Group so that the
+// Adapter code, and eventually file_sd, knows how to deal with it.
+func (d *discovery) jsonToTargets(data []byte) ([]*targetgroup.Group, error) {
```
For clarity you should probably inline this.

brian-brazil commented on documentation/examples/custom-sd/adapter-usage/main.go:
```diff
@@ -0,0 +1,156 @@
+		return
+	}
+
+	if *logFile != "" {
```
This all should go. We don't offer logging to file, and it's not relevant to the example.

brian-brazil commented on documentation/examples/custom-sd/adapter/adapter.go:
```diff
@@ -0,0 +1,148 @@
+	return arr
+}
+
+// Parses incomming target groups updates. If the update contains changes to the target groups
```
incoming

brian-brazil commented on documentation/examples/custom-sd/adapter/adapter.go:
```diff
@@ -0,0 +1,148 @@
+	arr := mapToArray(a.groups)
+	b, _ := json.MarshalIndent(arr, "", "    ")
+
+	tmpfile, err := ioutil.TempFile("", "sd-adapter")
```
You need to make sure this is in the dame directory as the output file, otherwise rename can fail if it's on a different filesystem.

brian-brazil commented on documentation/examples/custom-sd/adapter/adapter.go:
```diff
@@ -0,0 +1,148 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
```
This should have a comment that you don't need to touch this file.

cstyan replied:
added

brian-brazil commented on documentation/examples/custom-sd/adapter-usage/main.go:
```diff
@@ -0,0 +1,156 @@
+
+var (
+	a             = kingpin.New("sd adapter usage", "Tool to generate file_sd target files for unimplemented SD mechanisms.")
+	outputFile    = a.Flag("output.file", "Output file for file_sd compatible file.").Default("/tmp/custom_sd.json").String()
```
Use the current directory

cstyan commented on documentation/examples/custom-sd/adapter-usage/main.go:
```diff
@@ -0,0 +1,291 @@
+
+			// If the service address is not empty it should be used instead of the node address
+			// since the service may be registered remotely through a different node
+			var addr string
```
in this example only the address is used as the target

cstyan commented on documentation/examples/custom-sd/adapter-usage/main.go:
```diff
@@ -0,0 +1,291 @@
+			}
+			// fmt.Println("addr: ", addr)
+			target := model.LabelSet{model.AddressLabel: model.LabelValue(addr)}
+			tgroup.Labels[addressLabel] = model.LabelValue(node.Address)
```
all the other info from the service/<name> query is treated as labels

cstyan commented:
@brian-brazil sorry for the delay. 

I've modified the adapter usage example to have a simplified version of the consul discovery with some modifications. It watches all services known to consul (except consul itself), treats the address of the service as the target (see https://github.com/prometheus/prometheus/pull/3720/files#r171992290), and all the other info from the service/<name> query is treated as labels (see https://github.com/prometheus/prometheus/pull/3720/files#r171992330)

brian-brazil commented on documentation/examples/custom-sd/adapter-usage/main.go:
```diff
@@ -0,0 +1,291 @@
+	logger        log.Logger
+
+	// serviceLabel is the name of the label containing the service name.
+	serviceLabel = model.LabelName("consul_service")
```
All labels should be prefixed with __meta

cstyan replied:
I removed the `__meta` prefix since the labels will just end up in the file_sd config and then (I assume) be prepended with `__meta` again? If that's not the case I'll add them back with a comment.

brian-brazil replied:
No, the labels here are the labels. For a non-generic SD you wouldn't want to have to add extra relabelling.

brian-brazil commented on documentation/examples/custom-sd/adapter-usage/main.go:
```diff
@@ -0,0 +1,291 @@
+	watchTimeout  = 30 * time.Second
+	retryInterval = 15 * time.Second
+	// DefaultSDConfig is the default Consul SD configuration.
+	defaultSDConfig = sdConfig{
```
I'd hardcode all of this, it's not relevant to explaining how the interface works

brian-brazil commented on documentation/examples/custom-sd/adapter-usage/main.go:
```diff
@@ -0,0 +1,291 @@
+
+	var lastIndex uint64
+	for {
+		catalog := d.client.Catalog()
```
To keep things as simple as possible I'd not use the client library or watches, and instead use the Go http client directly with a simple refresh every say 30s. Most SDs seem to be of the refresh type rather than the watch type, and if someone needs and example of watching they can always look at other SD implementations.

cstyan commented:
hmmm, looks like the fetch of `staticcheck` is failing :|

cstyan commented:
@brian-brazil rebased down to one commit. Not sure what's happening with the travis `1.x` release, it seems to be using `1.10` same as the other build, which passes. All the tests pass for me locally on `1.10` as well.

brian-brazil commented on documentation/examples/custom-sd/adapter-usage/main.go:
```diff
@@ -0,0 +1,239 @@
+	// serviceIDLabel is the name of the label containing the service ID.
+	serviceIDLabel = model.MetaLabelPrefix + "consul_service_id"
+
+	watchTimeout  = 30 * time.Second
```
These are unused now

brian-brazil commented on documentation/examples/custom-sd/adapter-usage/main.go:
```diff
@@ -0,0 +1,239 @@
+	retryInterval = 15 * time.Second
+)
+
+// CatalogService is copied from https://github.com/hashicorp/consul/blob/master/api/catalog.go
```
If you're copying code from elsewhere you should include a link to the license

brian-brazil commented on documentation/examples/custom-sd/adapter-usage/main.go:
```diff
@@ -0,0 +1,239 @@
+	refreshInterval  int
+	clientDatacenter string
+	tagSeparator     string
+	watchedServices  []string // Set of services which will be discovered.
```
This is unused

brian-brazil commented on documentation/examples/custom-sd/adapter-usage/main.go:
```diff
@@ -0,0 +1,239 @@
+			}
+			resp, err := http.Get(fmt.Sprintf("http://%s/v1/catalog/service/%s", d.address, name))
+			if err != nil {
+				level.Error(d.logger).Log("msg", "Error getting services nodes", "service", name, "err", err)
```
You need to abort this iteration when this fails

brian-brazil commented on documentation/examples/custom-sd/adapter-usage/main.go:
```diff
@@ -0,0 +1,239 @@
+			if err != nil {
+				level.Error(d.logger).Log("msg", "Error getting services nodes", "service", name, "err", err)
+			}
+			ch <- d.parseServiceNodes(resp, name)
```
What if a service is removed? It might be easiest to have a single output target group with everything

brian-brazil commented on documentation/examples/custom-sd/adapter-usage/main.go:
```diff
@@ -0,0 +1,239 @@
+			}
+			ch <- d.parseServiceNodes(resp, name)
+		}
+		// wait for ticker or exit when ctx is closed
```
Comments should be full sentances

cstyan commented:
updated, again :)

brian-brazil commented on documentation/examples/custom-sd/adapter-usage/main.go:
```diff
@@ -0,0 +1,243 @@
+		// Add all key/value pairs from the node's metadata as their own labels.
+		for k, v := range node.NodeMeta {
+			name := strutil.SanitizeLabelName(k)
+			tgroup.Labels[model.LabelName(name)] = model.LabelValue(v)
```
Are you missing a label prefix here?

cstyan replied:
yep!

brian-brazil commented on documentation/examples/custom-sd/adapter-usage/main.go:
```diff
@@ -0,0 +1,243 @@
+			resp, err := http.Get(fmt.Sprintf("http://%s/v1/catalog/service/%s", d.address, name))
+			if err != nil {
+				level.Error(d.logger).Log("msg", "Error getting services nodes", "service", name, "err", err)
+				continue
```
This should abort this cycle, not continue on to the next service.

cstyan replied:
how come? we want to assume all `service/<serviceName>` endpoints will be down if one fails?

I was thinking that between the time we query for all services and the time we query for each services nodes that a service may be removed, so querying it's endpoint would fail but the rest may succeed

brian-brazil replied:
It'd better to have stale data than incomplete data.

cstyan replied:
In this case would it really be incomplete? A service is no longer available

brian-brazil replied:
How would we distinguish a service no longer existing from a timeout?

brian-brazil commented on documentation/examples/custom-sd/adapter-usage/main.go:
```diff
@@ -0,0 +1,243 @@
+			}
+			tg, err := d.parseServiceNodes(resp, name)
+			if err != nil {
+				level.Error(d.logger).Log("msg", "Error parsing services nodes", "name", name, "err", err)
```
This should also abort

brian-brazil commented on documentation/examples/custom-sd/adapter-usage/main.go:
```diff
@@ -0,0 +1,243 @@
+			}
+			tgs = append(tgs, tg)
+		}
+		ch <- tgs
```
I'd add a comment here that you're sending everything as one targetgroup

cstyan commented:
@brian-brazil okay, I think the loops for `Run` are correct now, I also added a comment to explain why we want to abort that iteration of the loop if there's an error when we're querying some service in Consul.

brian-brazil commented on documentation/examples/custom-sd/adapter-usage/main.go:
```diff
@@ -0,0 +1,251 @@
+			resp, err := http.Get(fmt.Sprintf("http://%s/v1/catalog/service/%s", d.address, name))
+			if err != nil {
+				level.Error(d.logger).Log("msg", "Error getting services nodes", "service", name, "err", err)
+				goto OUTER
```
I think you meant continue rather than goto

cstyan replied:
no I meant `goto`, but the way I'd set up the ticker this won't work

the label for the goto should be at the select statement below

brian-brazil replied:
I'd really like to avoid having a goto.

If the for is just being used for the initialiser, Id say it'd be cleaner to move the initialisation before the for.

brian-brazil replied:
I think the thing to do is move the err outside of the inner for, break the inner for if there's an error, and only send to the channel if there is no error.

caarlos0 commented on documentation/examples/custom-sd/adapter-usage/main.go:
```diff
@@ -0,0 +1,252 @@
+	a.Parse(os.Args[1:])
+	a.HelpFlag.Short('h')
+
+	_, err := a.Parse(os.Args[1:])
```
you're calling a.Parse 2 times: here and on line 224.

cstyan replied:
you're right, thanks!

cstyan commented:
@brian-brazil @krasi-georgiev PTAL :)

brian-brazil commented:
Thanks!
